index.js

```diff
@@ -155,45 +104,72 @@ function lint(input, config, webpack) {
  * @return {void}
  */
 module.exports = function(input, map) {
+  var webpack = this
   var config = assign(
     // loader defaults
     {
       formatter: require("eslint/lib/formatters/stylish"),
+      cacheIdentifier: JSON.stringify({
+        "eslint-loader": pkg.version,
+        eslint: eslint.version,
+      }),
     },
     // user defaults
     this.options.eslint || {},
     // loader query string
     loaderUtils.getOptions(this)
   )
-  this.cacheable()
+
+  var cacheDirectory = config.cacheDirectory
+  var cacheIdentifier = config.cacheIdentifier
+
+  delete config.cacheDirectory
+  delete config.cacheIdentifier
 
   // Create the engine only once per config
   var configHash = objectHash(config)
   if (!engines[configHash]) {
     engines[configHash] = new eslint.CLIEngine(config)
   }
 
-  // Read the cached information only once and if enable
-  if (cache === null) {
-    if (config.cache) {
-      var thunk = findCacheDir({
-        name: "eslint-loader",
-        thunk: true,
-        create: true,
-      })
-      cachePath = thunk("data.json") || os.tmpdir() + "/data.json"
-      try {
-        cache = require(cachePath)
-      }
-      catch (e) {
-        cache = {}
+  this.cacheable()
+
+  var resourcePath = webpack.resourcePath
+  var cwd = process.cwd()
+
+  // remove cwd from resource path in case webpack has been started from project
+  // root, to allow having relative paths in .eslintignore
+  if (resourcePath.indexOf(cwd) === 0) {
+    resourcePath = resourcePath.substr(cwd.length + 1)
+  }
+
+  var engine = engines[configHash]
+  // return early if cached
+  if (config.cache) {
+    var callback = this.async()
+    return cache(
+      {
+        directory: cacheDirectory,
+        identifier: cacheIdentifier,
+        options: config,
+        source: input,
+        transform: function() {
+          return lint(engine, input, resourcePath)
+        },
+      },
+      function(err, res) {
+        if (err) {
+          return callback(err)
+        }
+        printLinterOutput(res || {}, config, webpack)
+        return callback(null, input, map)
       }
-    }
-    else {
-      cache = false
-    }
+    )
   }
+  printLinterOutput(lint(engine, input, resourcePath), config, this)
+  this.callback(input, map)
```

Comment:
@viankakrisna i think this is your problem - first argument to `this.callback` is an error so should be set to `null` like the other `callback` a couple of line above it.